clients/src/main/java/org/apache/kafka/common/protocol/ApiKeys.java

```diff
@@ -127,97 +37,66 @@
-    DELETE_GROUPS(42, "DeleteGroups", DeleteGroupsRequestData.SCHEMAS, DeleteGroupsResponseData.SCHEMAS),
-    ELECT_LEADERS(43, "ElectLeaders", ElectLeadersRequestData.SCHEMAS,
-            ElectLeadersResponseData.SCHEMAS),
-    INCREMENTAL_ALTER_CONFIGS(44, "IncrementalAlterConfigs", IncrementalAlterConfigsRequestData.SCHEMAS,
-                              IncrementalAlterConfigsResponseData.SCHEMAS),
-    ALTER_PARTITION_REASSIGNMENTS(45, "AlterPartitionReassignments", AlterPartitionReassignmentsRequestData.SCHEMAS,
-                                  AlterPartitionReassignmentsResponseData.SCHEMAS),
-    LIST_PARTITION_REASSIGNMENTS(46, "ListPartitionReassignments", ListPartitionReassignmentsRequestData.SCHEMAS,
-                                 ListPartitionReassignmentsResponseData.SCHEMAS);
-
-    private static final ApiKeys[] ID_TO_TYPE;
-    private static final int MIN_API_KEY = 0;
-    public static final int MAX_API_KEY;
+    PRODUCE(0),
+    FETCH(1),
+    LIST_OFFSETS(2),
+    METADATA(3),
+    LEADER_AND_ISR(4),
+    STOP_REPLICA(5),
+    UPDATE_METADATA(6),
+    CONTROLLED_SHUTDOWN(7),
+    OFFSET_COMMIT(8),
+    OFFSET_FETCH(9),
+    FIND_COORDINATOR(10),
+    JOIN_GROUP(11),
+    HEARTBEAT(12),
+    LEAVE_GROUP(13),
+    SYNC_GROUP(14),
+    DESCRIBE_GROUPS(15),
+    LIST_GROUPS(16),
+    SASL_HANDSHAKE(17),
+    API_VERSIONS(18),
+    CREATE_TOPICS(19),
+    DELETE_TOPICS(20),
+    DELETE_RECORDS(21),
+    INIT_PRODUCER_ID(22),
+    OFFSET_FOR_LEADER_EPOCH(23),
+    ADD_PARTITIONS_TO_TXN(24),
+    ADD_OFFSETS_TO_TXN(25),
+    END_TXN(26, false, RecordBatch.MAGIC_VALUE_V2),
+    WRITE_TXN_MARKERS(27, true, RecordBatch.MAGIC_VALUE_V2),
+    TXN_OFFSET_COMMIT(28, false, RecordBatch.MAGIC_VALUE_V2),
+    DESCRIBE_ACLS(29),
+    CREATE_ACLS(30),
+    DELETE_ACLS(31),
+    DESCRIBE_CONFIGS(32),
+    ALTER_CONFIGS(33),
+    ALTER_REPLICA_LOG_DIRS(34),
+    DESCRIBE_LOG_DIRS(35),
+    SASL_AUTHENTICATE(36),
+    CREATE_PARTITIONS(37),
+    CREATE_DELEGATION_TOKEN(38),
+    RENEW_DELEGATION_TOKEN(39),
+    EXPIRE_DELEGATION_TOKEN(40),
+    DESCRIBE_DELEGATION_TOKEN(41),
+    DELETE_GROUPS(42),
+    ELECT_LEADERS(43),
+    INCREMENTAL_ALTER_CONFIGS(44),
+    ALTER_PARTITION_REASSIGNMENTS(45),
+    LIST_PARTITION_REASSIGNMENTS(46);
+
+    private static final Map<Integer, ApiKeys> ID_TO_API_KEY;
 
     static {
-        int maxKey = -1;
-        for (ApiKeys key : ApiKeys.values())
-            maxKey = Math.max(maxKey, key.id);
-        ApiKeys[] idToType = new ApiKeys[maxKey + 1];
-        for (ApiKeys key : ApiKeys.values())
-            idToType[key.id] = key;
-        ID_TO_TYPE = idToType;
-        MAX_API_KEY = maxKey;
+        HashMap<Integer, ApiKeys> map = new HashMap<>();
+        for (ApiKeys key : ApiKeys.values()) {
+            map.put(Integer.valueOf(key.id), key);
+        }
```

Comment:
Why did we change it from an array to a HashMap?

Reply:
I guess my thinking was that it's more hassle to deal with bounds checking on the array than to call `map#get`, and they're both O(1).